File: JAVA/Hashing/_169_MajorityElement.java
// Problem: https://leetcode.com/problems/majority-element/
// Note: This is a better solution after the brute force one. For the most optimal solution, refer "../Arrays/_169_MajorityElement.java"

/* Solution:
 * Initialize the hashmap and store the count of each element in the array.
 * Iterate through the hashmap and return the element with the count greater than n/2
*/

import java.util.HashMap;

public class _169_MajorityElement{
    private static int majorityElement(int[] nums){
        HashMap<Integer,Integer> map = new HashMap<Integer,Integer>();
        for(int i=0; i < nums.length; i++){
            if(map.get(nums[i]) == null){
                map.put(nums[i], 1);
            }
            else{
                map.put(nums[i], map.get(nums[i]) + 1);
            }
        }
        for(int major : map.keySet()){
            if(map.get(major) > nums.length/2){
                return major;
            }
        }
        return 0;
    }
    public static void main(String[] args){
        int[] nums = {2,2,1,1,1,2,2};
        System.out.println(majorityElement(nums));
    }
}

// Time Complexity: O(n)  --> Taken for iterating through the array and simultaneously storing the count of each element
// Space Complexity: O(n)  --> Taken for storing the elements in the hashmap

File: C++/Hashing/_169_MajorityElement.cpp
// Problem: https://leetcode.com/problems/majority-element/
// Note: This is a better solution after the brute force one. For the most optimal solution, refer "../Arrays/_169_MajorityElement.java"

/* Solution:
 * Initialize the hashmap and store the count of each element in the array.
 * Iterate through the hashmap and return the element with the count greater than n/2
*/

#include <bits/stdc++.h>
#include <vector>
#include <map>
using namespace std;

int majorityElement(vector<int> nums){
    map<int,int> counter;
    for(int i=0; i < nums.size(); i++){
        counter[nums[i]]++;
    }
    for(auto major : counter){
        if(major.second > nums.size()/2){
            return major.first;
        }
    }
    return -1;
}

int main(){
    vector<int> nums = {2,2,1,1,1,2,2};
    cout << majorityElement(nums) << endl;
}


// Time Complexity: O(n)  --> Taken for iterating through the array and simultaneously storing the count of each element
// Space Complexity: O(n)  --> Taken for storing the elements in the hashmap